operators/src/processing/time_shift.rs

```diff
@@ -24,25 +30,106 @@ pub type TimeShift = Operator<TimeShiftParams, SingleRasterOrVectorSource>;
 pub enum TimeShiftParams {
     /// Shift the query rectangle relative with a time step
     Relative {
-        step: TimeStep,
-        direction: RelativeShiftDirection,
+        granularity: TimeGranularity,
+        value: i32,
     },
     /// Set the time interval to a fixed value
     Absolute { time_interval: TimeInterval },
 }
 
-#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Copy)]
-#[serde(rename_all = "camelCase")]
-pub enum RelativeShiftDirection {
-    Forward,
-    Backward,
+pub trait TimeShiftOperation: Send + Sync + Copy {
+    type State: Send + Sync + Copy;
+
+    fn shift(&self, time_interval: TimeInterval) -> Result<(TimeInterval, Self::State)>;
```

Comment:
wouldn't if suffice short `shift` to return the amount shifted? Would we ever need to return something else?

Yes, not all shifts need the amount shifted for calculating the reverse shift, but I guess the compiler would optimize that.

It would make the code more easily understandable, but of course it's already fine the way it is. What do you think?

Reply:
Well, you need the new time interval as a result at first.
Then, you could always return the two durations shifted. But doing it this way we can omit that computation and storage for the relative shift. And the union state is definitely optimized out.

So, I would leave it as is. It is not 100% nice design, but works and it is just one state var.